bpo-27377: Add socket.fdtype()

nascheme opened the pull request:
This is a recreation of pull request #1333 as there seems no way to move the source branch into my own repo.  I have removed the redundant definitions as pointed out by @tiran.  I could maybe be convinced that socket.socket() get this functionality as suggested.

Comment from @tiran copied here::

>   I'm -1 on fromfd2. We shouldn't introduce yet another function and bloat the API even further. Instead 
>  let's fix the underlying bug and make socket.socket(fileno=fd) detect the socket values, see 
>  https://bugs.python.org/issue28134

DimitrisJim commented on Doc/library/socket.rst:
```diff
@@ -506,6 +506,22 @@ The following functions all create :ref:`socket objects <socket-objects>`.
+
+   Availability: Unix.
+
+   .. versionadded:: 3.6
```
Shouldn't the `versionadded` (and similarly for `fdtype`) be `3.7`?

tiran commented:
#1349 implements a fix that does not require a new API and fixes all existing usage of socket(fileno) automatically.

nascheme committed:
Add socket.fdtype().

Get socket information from a file descriptor.  The return value is a
3-tuple with the following structure: (family, type, proto).  Not
supported on all platforms.

nascheme commented:
Changed title, I think with the socket.socket change proposed by @tiran, we don't need fromf2().  I still think socket.fdtype() is potentially useful.

As suggested by @DimitrisJim, I have fixed the version number to be 3.7.

The HAVE_FDTYPE define may need some tweaking yet.  Based on testing I did some months ago, platforms define SO_TYPE and SO_PROTOCOL but the fdtype() implementation doesn't work on them.

Having this function in socket would allow the unit tests for #1349 to check for it and only run the tests if fdtype() is supported on the platform.

tiran commented on Modules/socketmodule.c:
```diff
@@ -5470,6 +5470,61 @@ AF_UNIX if defined on the platform; otherwise, the default is AF_INET.");
+    }
+
+    l = sizeof(sa);
+    if (getsockname(fd, &sa, &l) < 0) {
```
Your code doesn't work on Windows for IPv6 addresses. You have to make the buffer large enough to fit all possible address types in it.

```
sock_addr_t addrbuf;
socklen_t addrlen = sizeof(sock_addr_t);

memset(&addrbuf, 0, addrlen);
if (getsockname(fd, SAS2SA(&addrbuf), &addrlen) == 0) {
    family = SAS2SA(&addrbuf)->sa_family;
} else {
    return set_error();
}
```
